router/proxy/transport.go

```diff
@@ -64,13 +66,33 @@ func (t *transport) setStickyBackend(res *http.Response, originalStickyBackend s
 	}
 }
 
-// always sets the response.Request.URL.Host to the last backend that was
-// connected to.
-func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+func (t *transport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
 	// http.Transport closes the request body on a failed dial, issue #875
 	req.Body = &fakeCloseReadCloser{req.Body}
 	defer req.Body.(*fakeCloseReadCloser).RealClose()
 
+	resultc := make(chan httpResult, 1)
+
+	go func() {
+		defer close(resultc)
+		res, err := t.roundTrip(req)
+		resultc <- httpResult{res, err}
+	}()
+
+	select {
+	case <-ctx.Done():
+		httpTransport.CancelRequest(req)
```

Comment:
Another subtle case to handle is covered in the review for that `httputil.ReverseProxy` change: https://go-review.googlesource.com/#/c/2320/5/src/net/http/httputil/reverseproxy.go@120

> If the CloseNotifier triggers very quickly (like if the client sends a request and then immediately closes their connection), it seems like this call could happen before the call to transport.RoundTrip begins.

So that's why they don't spawn their cancellation goroutine until after the Transport starts reading the request body.